// @astrojs/cloudflare/dist/wrangler.js
const DEFAULT_SESSION_KV_BINDING_NAME = "SESSION";
const DEFAULT_IMAGES_BINDING_NAME = "IMAGES";
const DEFAULT_ASSETS_BINDING_NAME = "ASSETS";
function cloudflareConfigCustomizer(options) {
  const sessionKVBindingName = options?.sessionKVBindingName ?? DEFAULT_SESSION_KV_BINDING_NAME;
  const needsSessionKVBinding = options?.needsSessionKVBinding ?? true;
  const imagesBindingName = options?.imagesBindingName === false ? void 0 : options?.imagesBindingName ?? DEFAULT_IMAGES_BINDING_NAME;
  return (config) => {
    const hasSessionBinding = config.kv_namespaces?.some(
      (kv) => kv.binding === sessionKVBindingName
    );
    const hasImagesBinding = config.images?.binding !== void 0;
    const hasAssetsBinding = config.assets?.binding !== void 0;
    return {
      main: config.main ?? "@astrojs/cloudflare/entrypoints/server",
      kv_namespaces: !needsSessionKVBinding || hasSessionBinding ? void 0 : [
        {
          binding: sessionKVBindingName
        }
      ],
      images: hasImagesBinding || !imagesBindingName ? void 0 : {
        binding: imagesBindingName
      },
      assets: hasAssetsBinding ? void 0 : {
        binding: DEFAULT_ASSETS_BINDING_NAME
      }
    };
  };
}
export {
  DEFAULT_ASSETS_BINDING_NAME,
  DEFAULT_IMAGES_BINDING_NAME,
  DEFAULT_SESSION_KV_BINDING_NAME,
  cloudflareConfigCustomizer
};
