// @keystar/ui/dist/keystar-ui-core.js
import { u as useProvider, d as documentElementClasses } from './KeystarProvider-dac0e778.js';
export { K as KeystarProvider, K as TestProvider, d as documentElementClasses, u as useProvider, a as useProviderProps } from './KeystarProvider-dac0e778.js';
import { c } from './react-compiler-runtime-19061925.js';
import { useLayoutEffect } from 'react';
import '@react-aria/i18n';
import '@react-aria/overlays';
import '@react-aria/utils';
import '@keystar/ui/style';
import '@keystar/ui/utils/ts';
import '@keystar/ui/primitives';
import 'react/jsx-runtime';

function ClientSideOnlyDocumentElement(props) {
  const $ = c(6);
  const context = useProvider();
  let t0;
  if ($[0] !== props.bodyBackground || $[1] !== context.colorScheme) {
    t0 = documentElementClasses({
      bodyBackground: props.bodyBackground,
      colorScheme: context.colorScheme
    });
    $[0] = props.bodyBackground;
    $[1] = context.colorScheme;
    $[2] = t0;
  } else {
    t0 = $[2];
  }
  const classes = t0;
  let t1;
  let t2;
  if ($[3] !== classes) {
    t1 = () => {
      const split = classes.split(" ");
      const root = document.documentElement;
      root.classList.add(...split);
      return () => {
        root.classList.remove(...split);
      };
    };
    t2 = [classes];
    $[3] = classes;
    $[4] = t1;
    $[5] = t2;
  } else {
    t1 = $[4];
    t2 = $[5];
  }
  useLayoutEffect(t1, t2);
  return null;
}

export { ClientSideOnlyDocumentElement };
